tests/SpecTests/ClientSideEncryptionSpecTest.php

```diff
@@ -0,0 +1,757 @@
+            $jsonSchema = isset($json->json_schema) ? $json->json_schema : null;
+            $databaseName = isset($json->database_name) ? $json->database_name : null;
+            $collectionName = isset($json->collection_name) ? $json->collection_name : null;
+
+            foreach ($json->tests as $test) {
+                $name = $group . ': ' . $test->description;
+                $testArgs[$name] = [$test, $runOn, $data, $keyVaultData, $jsonSchema, $databaseName, $collectionName];
+            }
+        }
+
+        return $testArgs;
+    }
+
+    /**
+     * @dataProvider dataKeyProvider
+     */
+    public function testDataKeyAndDoubleEncryption(Closure $test)
+    {
+        $this->skipIfClientSideEncryptionIsNotSupported();
+
+        $client = new Client(static::getUri());
+
+        $client->selectCollection('admin', 'datakeys')->drop();
+        $client->selectCollection('db', 'coll')->drop();
+
+        $encryptionOpts = [
+            'keyVaultNamespace' => 'admin.datakeys',
+            'kmsProviders' => [
+                'aws' => $this->getAWSCredentials(),
+                'local' => ['key' => new Binary(base64_decode(self::LOCAL_MASTERKEY), 0)],
+            ],
+        ];
+
+        $autoEncryptionOpts = $encryptionOpts + [
+            'schemaMap' => [
+                'db.coll' => [
+                    'bsonType' => 'object',
+                    'properties' => [
+                        'encrypted_placeholder' => [
+                            'encrypt' => [
+                                'keyId' => '/placeholder',
+                                'bsonType' => 'string',
+                                'algorithm' => 'AEAD_AES_256_CBC_HMAC_SHA_512-Random',
+                            ],
+                        ],
+                    ],
+                ],
+            ],
+            'keyVaultClient' => $client,
+        ];
+
+        $clientEncrypted = new Client(static::getUri(), [], ['autoEncryption' => $autoEncryptionOpts]);
+        $clientEncryption = $clientEncrypted->createClientEncryption($encryptionOpts);
+
+        $test($clientEncryption, $client, $clientEncrypted, $this);
+    }
+
+    public function dataKeyProvider()
+    {
+        return [
+            'local' => [
+                function (ClientEncryption $clientEncryption, Client $client, Client $clientEncrypted, self $test) {
+                    $commands = [];
+                    $localDatakeyId = null;
+
+                    (new CommandObserver())->observe(
+                        function () use ($clientEncryption, &$localDatakeyId) {
+                            $localDatakeyId = $clientEncryption->createDataKey('local', ['keyAltNames' => ['local_altname']]);
+                        },
+                        function ($command) use (&$commands) {
+                            $commands[] = $command;
+                        }
+                    );
+
+                    $this->assertInstanceOf(Binary::class, $localDatakeyId);
```

Comment:
This is a bit tricky. `$this` is the instance of the test that was used to call the data provider. However, PHPUnit then uses a separate instance to run the actual tests, which is why we can't use `$this` to expect an exception. To remove the ambiguity, I made the closure static to avoid binding `$this` to its scope and always used the passed `$test` instance, even for assertions. Last but not least, I made the data provider itself static (which PHPUnit supports) to further remove issues due to using `$this`.

Reply:
TIL about [static anonymous functions](https://www.php.net/manual/en/functions.anonymous.php#functions.anonymous-functions.static) :smile: